File: haskell/secret-handshake/src/SecretHandshake.hs
module SecretHandshake (handshake) where
import Data.Bits (testBit)

handshake :: Int -> [String]
handshake n = (if (testBit n 4) then reverse else id)
              $ map fst
              $ filter (\pair -> testBit n $ snd pair)
              $ zip ["wink","double blink","close your eyes","jump"] [0..]

File: fsharp/secret-handshake/SecretHandshake.fs
module SecretHandshake

let codes =
    [ (1, "wink")
      (2, "double blink")
      (4, "close your eyes")
      (8, "jump") ]

let commands num =
    let r =
        codes
         |> List.filter (fun (code, _) -> num &&& code > 0)
         |> List.map snd
    if num &&& 16 = 16 then List.rev r else r
